### PR TITLE
[dmd-cxx] Backport static foreach optimization

### PR DESCRIPTION
**Pull request by BorisCarvajal**
ping @ibuclaw

### CODE REVIEW

File: src/cond.c [hunk 1]
```diff
@@ -307,13 +318,50 @@ static void lowerNonArrayAggregate(StaticForeach *sfe, Scope *sc)
...
+
+        if (sfe->rangefe->op == TOKforeach)
+        {
+            for (size_t i = 0; i < length; i++)
+                (*exps)[i] = new IntegerExp(aloc, lwr + i, indexty);
+        }
+        else
+        {
+            --upr;
+            for (size_t i = 0; i < length; i++)
```
Comment (ibuclaw) [inline src/cond.c:350]: This should be `foreach_reverse`
```suggestion
            for (size_t i = length;  0 < i--; )
```

File: src/cond.c [hunk 2]
```diff
@@ -90,15 +90,26 @@ static void lowerArrayAggregate(StaticForeach *sfe, Scope *sc)
     sc = sc->endCTFE();
     el = el->optimize(WANTvalue);
     el = el->ctfeInterpret();
+    Expressions *es;
     if (el->op == TOKint64)
     {
-        dinteger_t length = el->toInteger();
-        Expressions *es = new Expressions();
-        for (size_t i = 0; i < length; i++)
+        if (aggr->isArrayLiteralExp())
```
Comment (ibuclaw) [inline src/cond.c:96]:
```suggestion
        if (ArrayLiteralExp *ale = aggr->isArrayLiteralExp())
```

File: src/cond.c [hunk 3]
```diff
@@ -90,15 +90,26 @@ static void lowerArrayAggregate(StaticForeach *sfe, Scope *sc)
     sc = sc->endCTFE();
     el = el->optimize(WANTvalue);
     el = el->ctfeInterpret();
+    Expressions *es;
     if (el->op == TOKint64)
     {
```
Comment (ibuclaw) [inline src/cond.c:95]:
```suggestion
    if (el->op == TOKint64)
    {
        Expressions *es;
```

File: src/cond.c (hunk 1 above)
Comment (BorisCarvajal) [reply src/cond.c:350]: @ibuclaw , I changed the original back to `foreach`, it doesn't matter. Feels more natural to fill the elements from left to right.

Comment (ibuclaw) [reply src/cond.c:349]: OK.

**Comment by ibuclaw** [comment]
Regression testing...

**Pull request merged**